sanic/errorpages.py

```diff
@@ -398,34 +401,89 @@ def exception_response(
-                # we can assume that's what we want. Otherwise, go back to
-                # previous guess.
-                renderer = RENDERERS_BY_CONTENT_TYPE.get(
-                    content_type, renderer
-                )
+                # First, if there is an Accept header, check if text/html
+                # is the first option
+                # According to MDN Web Docs, all major browsers use text/html
+                # as the primary value in Accept (with the exception of IE 8,
+                # and, well, if you are supporting IE 8, then you have bigger
+                # problems to concern yourself with than what default exception
+                # renderer is used)
+                # Source:
+                # https://developer.mozilla.org/en-US/docs/Web/HTTP/Content_negotiation/List_of_default_Accept_values
+
+                if acceptable and acceptable[0].match(
+                    "text/html",
+                    allow_type_wildcard=False,
+                    allow_subtype_wildcard=False,
+                ):
+                    renderer = HTMLRenderer
+
+                # Second, if there is an Accept header, check if
+                # application/json is an option, or if the content-type
+                # is application/json
+                elif (
+                    acceptable
+                    and acceptable.match(
+                        "application/json",
+                        allow_type_wildcard=False,
+                        allow_subtype_wildcard=False,
+                    )
+                    or content_type == "application/json"
+                ):
+                    renderer = JSONRenderer
+
+                # Third, if there is no Accept header, assume we want text.
+                # The likely use case here is a raw socket.
+                elif not acceptable:
+                    renderer = TextRenderer
+                else:
+                    # If the client is curl, we will default to TextRenderer
+                    default = (
+                        TextRenderer
+                        if "curl"
+                        in request.headers.get("user-agent", "").lower()
+                        else base
+                    )
+
+                    # Fourth, look to see if there was a JSON body
+                    # When in this situation, the request is probably coming
+                    # from curl, an API client like Postman or Insomnia, or a
+                    # package like requests or httpx
+                    try:
+                        # Give them the benefit of the doubt if they did:
+                        # $ curl localhost:8000 -d '{"foo": "bar"}'
+                        # And provide them with JSONRenderer
+                        renderer = (
+                            JSONRenderer  # type: ignore
+                            if request.json
+                            else default
+                        )
```

Comment:
I would leave these out. Custom handling for curl can be confusing when debugging why it doesn't work in browser or elsewhere.

Reply:
Good point

Reply:
Okay, I guess this gets simplified at 22.3. Maybe emit warnings for the cases that would be changed then or in case for the json body detection, could be changed maybe a year later? (reading body to determine type is problematic and just evil).